core/src/poh_recorder.rs

```diff
@@ -135,39 +127,41 @@ impl PohRecorder {
 
     /// returns if leader tick has reached, how many grace ticks were afforded,
     ///   imputed leader_slot and self.start_slot
+    /// reached_leader_tick() == true means "ready for a bank"
     pub fn reached_leader_tick(&self) -> (bool, u64, Slot, Slot) {
-        let slot =
-            leader_schedule_utils::tick_height_to_slot(self.ticks_per_slot, self.tick_height);
-
         trace!(
-            "tick_height {}, start_tick {} start_leader_at_tick {:?}, grace {}",
+            "tick_height {}, start_tick {}, start_leader_at_tick {:?}, grace {}, has_bank {}",
             self.tick_height,
             self.start_tick,
             self.start_leader_at_tick,
-            self.max_last_leader_grace_ticks
+            self.grace_ticks,
+            self.has_bank()
         );
 
+        let next_tick = self.tick_height + 1;
+
         if let Some(target_tick) = self.start_leader_at_tick {
-            let leader_ideal_start_tick =
-                target_tick.saturating_sub(self.max_last_leader_grace_ticks);
-            // Check if either grace period has expired,
-            // or target tick is = grace period (i.e. poh recorder was just reset)
-            if self.tick_height >= target_tick
-                || self.max_last_leader_grace_ticks >= target_tick.saturating_sub(self.start_tick)
+            // if we've reached the target tick OR poh was *just* reset
+            if next_tick >= target_tick
+                || next_tick >= self.start_tick && self.start_tick + self.grace_ticks == target_tick
             {
+                let ideal_target_tick = target_tick.saturating_sub(self.grace_ticks);
+
                 return (
                     true,
-                    self.tick_height.saturating_sub(leader_ideal_start_tick),
-                    slot,
+                    next_tick.saturating_sub(ideal_target_tick),
+                    next_tick / self.ticks_per_slot,
```

Comment:
should we use `leader_schedule_utils::tick_height_to_slot` here?

Reply:
no, that function should die IMHO